### ISSUE TITLE
Only invoke setup() once, not in both trainer.fit() and trainer.test() - #2620 follow up

### ISSUE THREAD

**Issue by AlexHarn** [PyTorchLightning/pytorch-lightning#9865]
## 🐛 Bug

It seems like the exact bug described in #2620 is back: When calling `trainer.test()` after `trainer.fit()`, `LightningDataModule.setup()` is called twice, which is especially problematic when using `random_split` in `setup()` because it leads to training samples leaking into the test set (in the worst case of extremely large training data compared to test data, the test data will most likely consist exclusively of training samples).

### To Reproduce

```Python
import os

import torch
from pytorch_lightning import LightningDataModule
from torch.utils.data import DataLoader, Dataset, random_split

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class MyDataModule(LightningDataModule):
    def __init__(self):
        super().__init__()

    def prepare_data(self):
        # download, split, etc...
        # only called on 1 GPU/TPU in distributed
        pass

    def setup(self, stage):
        # make assignments here (val/train/test split)
        # called on every process in DDP
        print("\nSetting up!!!\n")
        dataset = RandomDataset(32, 96)
        self.data_train, self.data_val, self.data_test = random_split(
            dataset, [32, 32, 32]
        )

    def train_dataloader(self):
        return DataLoader(self.data_train, batch_size=2)

    def val_dataloader(self):
        return DataLoader(self.data_val, batch_size=2)

    def test_dataloader(self):
        return DataLoader(self.data_test, batch_size=2)


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    datamodule = MyDataModule()

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        weights_summary=None,
    )
    trainer.fit(model=model, datamodule=datamodule)
    trainer.test(model=model, datamodule=datamodule)


if __name__ == "__main__":
    run()
```

### Expected behavior
I am expecting `MyDataModule.setup()` to be called once, but it is called twice (easily verified by the print statements).

### Environment

* CUDA:
        - GPU:
                - Tesla V100S-PCIE-32GB
                - Tesla V100S-PCIE-32GB
                - Tesla V100S-PCIE-32GB
                - Tesla V100S-PCIE-32GB
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.3
        - pyTorch_debug:     False
        - pyTorch_version:   1.9.1
        - pytorch-lightning: 1.4.8
        - tqdm:              4.60.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.11
        - version:           #1 SMP Wed Jul 21 11:57:15 UTC 2021

### Additional context

Maybe I misunderstand how `prepare()` and `setup()` are supposed to be used, especially in DDP mode, but the comment in the docs clearly states val/train/test split in `setup()`, which currently leads to the described problem of training data ending up in the test split when using a random split.

**Comment by AlexHarn** [comment]
I see, that makes a lot of sense to use it that way, thank you!

It just was not obvious to me from reading the docs/examples.
